```{r}
#PartI

#This program demonstrates how to read in a FASTA file and do some simple
#processing of the sequence
#
#A fasta file is a plain text file containing sequences
#a line begining with a ">"indicates a label
#all other lines below this one (that do not start with ">") is the sequence
#for that label
#
#

fileName = "seq.fasta"  #Change this to use your own file.

fileContents = read.table(fileName, sep="\t", header=T, stringsAsFactors=F, check.names=F)

sequence = paste(fileContents[[1]], sep="", collapse="")

cat ("Loaded sequence", names(fileContents),"\n")
cat ("Sequence:\n", sequence, "\n")

compSequence = chartr("ATCG", "TAGC", sequence)


cat ("The complement sequence is:\n", compSequence, "\n\n")

#split the string into an array of characters
sequenceVector = strsplit(compSequence, NULL)[[1]]
    
aCount = 0;
cCount = 0;
gCount = 0;
tCount = 0;
sequenceLength = 0;
for(base in sequenceVector)
{
	if(base == "A") {
		aCount = aCount+1
	}
	else if(base == "C") {
		cCount = cCount+1
	}
	else if(base == "G") {
		gCount = gCount+1
	}
	else if(base == "T") {
		tCount = tCount+1
	}
	sequenceLength = sequenceLength + 1
}
cat ("Sequence statistics:\n")
cat ("Fraction of A's : ",aCount/sequenceLength,"\n")
cat ("Fraction of T's : ",tCount/sequenceLength,"\n")
cat ("Fraction of G's : ",gCount/sequenceLength,"\n")
cat ("Fraction of C's : ",cCount/sequenceLength,"\n")
cat ("------------------------------\n")


```
```{r}
#Part II

# Script: buggy_script.r
# Description: Fix the bugs!
#

#Fix the errors.
cat ("this Seems to be \"ok\"\n")

sum = 0
sum = sum + 1
sum = sum + 1
sum = sum + 2
sum = sum + 2
cat (sum,"\n")

sequence1 = "AAAA"
sequence2 = "TTTT"
sequenceSum = paste0(sequence1,sequence2)  #correct this to give the proper concatenation.
print (sequenceSum)

cat ("\n-------------------\n")

emailattheU = "myemail.utoronto.ca"
cat (emailattheU,"\n")

sequence1 = "What's going on here?"
print (sequence1)

```
```{r}
#Part III

# Script: stringnumber_practice.r
# Description:  Fill in/answer all of the questions in this script

numbers = c(1, 2.5, 3.4, 7.8)
#1. Compute and print the mean and standard deviation of these four numbers
mean(numbers)
sd(numbers)
cat ("\n-------------------\n")

cat (numbers - 5, "\n")
cat (numbers[1:4] - 5)
#2. Explain what how the above two lines are different
#The first line substracts 5 from all elements of "numbers" vector, while the second line subtracts 5 only from first 4 elements, which happen to be the total number of elements in "numbers". 
cat ("\n-------------------\n")

sequence = "ATCAAATTCCGCGCATTCTTAGGCCATTCTT"
cat (sequence,"\n")
#3. Mask the subsequence "ATTC" with "xxxx"  and print the new sequence
sequence=gsub("ATTC","xxxx", sequence)
cat(sequence, "\n")
cat ("\n-------------------\n")
cat (sequence,"\n")
#4. Remove the x's from the sequence
cat(gsub("x","",sequence),"\n")
cat ("\n-------------------\n")

string = "This is a kind of a long string"
#5. Use the substr function to create a string with only the last word this sentence and print it out
substr(string,nchar(string)-5,nchar(string))
#you may try this: what happens when you put a negative number for the start index?
substr(string, -5, nchar(string))
#This works, but there are easier, more conventional ways of solving this problem.
tail(strsplit(string, " ")[[1]], n = 1)
cat ("\n-------------------\n")

```
```{r}
#Part IV

# Script: translate_sequence.r
# Problem description: read in sequence from file, compute stats, translate RNA into
# 	       amino acid sequence

#read sequence file into a variable
rna_sequence=read.table("assignment2_sequence.txt",sep="\t")
rna_sequence

#read 5 lines of the sequence into a string. Use this string for the rest of the code below.
r_sequence = paste(readLines("assignment2_sequence.txt", 5), sep="", collapse="")
r_sequence

#print the length of the sequence
nchar(r_sequence)

#get the coding part of the sequence (61st bp-->end) and store in variable called sequence
sequence=substr(r_sequence, 61, nchar(r_sequence))
sequence
#Fill in the loop below so that each codon is translated into the corresponding
#amino acid.  Append each codon to the end of the variable aasequence so that
#at the end of the loop, the variable aasequence contains the complete sequence
#of amino acids

aasequence = c()
   refTable = data.frame(cod = c("AUG", "UUC", "CUU", "CCA", "ACC",
                                   "GAU", "CGA", "UGA"),
                         aa = c("Start", "Phe", "Leu", "Pro", "Thr", "Asp",
                                "Arg", "Stop"))
   
index=1

aa2 = sapply(3*(0:((nchar(sequence) -1) / 3)) + 1, function(i) {
  codon = substr(sequence, i, i+ 2)
  subset(refTable, cod == codon, aa)
})
aa2 = paste0(as.character(unlist(aa2)), collapse = "")
while(index < nchar(sequence)) {

    # HINT: index contains the starting index of the current codon

    # get a substring corresponding to the next codon (3 bp)
   codon<-substr(sequence, index, index+2)
codon
    # Translate the codon into the corresponding amino acid.
   
    # Include the start and stop translations from the reference table.
   aasequence = c(aasequence, subset(refTable, cod == codon, aa))
   
    # Here's a reference table for the codons that appear in this sequence


    # concatenate the current codon onto the end of the aasequence variable

    index = index + 3  #this moves the index forward 3 places in the sequence
  }
  aasequence<-paste0(as.character(unlist(aasequence)), collapse = "")
   aasequence

#print the amino acid sequence to the output file "aaoutput.txt". Hint: ?cat
   cat(aasequence, file = "aasequence_output.txt")

```
```{r}
Solution to part IV


```

